cipher/cbc_hmac.go

```diff
@@ -23,15 +23,95 @@ import (
 	"crypto/sha256"
 	"crypto/sha512"
 	"crypto/subtle"
+	"encoding/base64"
 	"encoding/binary"
 	"errors"
 	"hash"
+	"strings"
 )
 
 const (
 	nonceBytes = 16
 )
 
+// ComputeIntegrityKey derives an integrity key based on a master key, using the A128CBC+HS256 draft specification
+func ComputeIntegrityKey(key []byte, algorithm string) []byte {
+	// content integrity key size is the
+	algBytes := []byte(algorithm)
+
+	buf := []byte{0, 0, 0, 1}
+	buf = append(buf, key...)
+
+	tmp := make([]byte, 4)
+	binary.BigEndian.PutUint32(tmp, uint32(len(key)*8))
+	buf = append(buf, tmp...)
+	buf = append(buf, algBytes...)
+	buf = append(buf, []byte{0, 0, 0, 0}...)
+	buf = append(buf, []byte{0, 0, 0, 0}...)
+	buf = append(buf, []byte("Integrity")...)
+
+	hashed := sha256.Sum256(buf)
+
+	return hashed[:len(key)]
+}
+
+// ComputeEncryptionKey derives an encryption key based on a master key, using the A128CBC+HS256 draft specification
```

Comment:
These could actually be made a little more generic to support A128CBC+HS384/512 too. You can take a hash.Hash as an argument and use that instead of calling `sha256.Sum256`. Everything else is portable.